README.md
# AWS Robo Advisor

This repo represents a simple Robo advisor created with AWS Lambda and AWS Lex. Based on user's answers, Robo adviser provides best portfolio for retirement plan.

## Approach

1. Create advisor bot using AWS Lex

2. Create and deploy lambda function using AWS Lambda

3. Test Robo advisor

---

## Technologies

This project leverages the following tools:

- [AWS Lambda](https://aws.amazon.com/lambda/) - is a serverless, event-driven compute service.

- [AWS Lex](https://aws.amazon.com/lex/) - to design, build, test, and deploy conversational interfaces in applications.

---

## Input data and how it works

Examples of how it works you can find in the folder `Video`.

Robo adviser asks questions and user provides answers:

Example:

```
User: I want to save money to my retirement
Bot: Thank you for trusting me to help. Couls you please give me your name?
User: John
Bot: How old are you?
User: 30
Bot: Ho much do you want to invest?
User: 10000
Bot: What level of investment risk would you like to take? (None, Low, Medium, High)
User: None
Bot: Thank you for your information, John. We recommend you an investment portfolio: 100% bonds (AGG) and 0% equities (SPY)

```

---

## Types of portfolios

```
None: “100% bonds (AGG), 0% equities (SPY)”
Low: “60% bonds (AGG), 40% equities (SPY)”
Medium: “40% bonds (AGG), 60% equities (SPY)”
High: “20% bonds (AGG), 80% equities (SPY)”
```

## Utterances

![Utterances](Images/utterances.JPG)

## Slots

![Slots](Images/slots.JPG)

## Test events

All test are passed. Pls, refer to `Video/Test_events.mp4`

---

## Contributors

Vladislav Glupak - [Linkedin](https://www.linkedin.com/in/vladislav-glupak/)

---

## License

MIT


history.txt
    1  . code
    2  conda activate dev
    3  git pull
    4  ls
    5  jupyter lab
    6  ls
    7  cd ..
    8  ls
    9  cd ..
   10  ls
   11  cd FinTech_modules/Module_15/
   12  cd ..
   13  ls
   14  git status
   15  git add .
   16  git commit -m "15 1 and 2"
   17  git push
   18  git status
   19  git add .
   20  git commit -m "15 1 and 2"
   21  git push
   22  git status
   23  git add .
   24  git commit -m "15 1 and 2"
   25  git push
   26  git status
   27  cd Module_15/
   28  ls
   29  https://github.com/VladislavGlupak/AWS-robo-advisor.git
   30  git clone https://github.com/VladislavGlupak/AWS-robo-advisor.git
   31  cd Starter_Code/
   32  git status
   33  git status
   34  git add .
   35  git commit -m "initial commit"
   36  git push
   37  git status
   38  git add .
   39  git commit -m "adding video, images, readme"
   40  git push
   41  git status
   42  git add .
   43  git commit -m "fix images links in the readme"
   44  git push
   45  git status
   46  git add .
   47  git commit -m "change license in readme"
   48  git push
   49  history 50 > history.txt


lambda_function.py
### Required Libraries ###
from datetime import datetime
from dateutil.relativedelta import relativedelta

### Functionality Helper Functions ###
def parse_int(n):
    """
    Securely converts a non-integer value to integer.
    """
    try:
        return int(n)
    except ValueError:
        return float("nan")


def build_validation_result(is_valid, violated_slot, message_content):
    """
    Define a result message structured as Lex response.
    """
    if message_content is None:
        return {"isValid": is_valid, "violatedSlot": violated_slot}

    return {
        "isValid": is_valid,
        "violatedSlot": violated_slot,
        "message": {"contentType": "PlainText", "content": message_content},
    }


### Dialog Actions Helper Functions ###
def get_slots(intent_request):
    """
    Fetch all the slots and their values from the current intent.
    """
    return intent_request["currentIntent"]["slots"]


def elicit_slot(session_attributes, intent_name, slots, slot_to_elicit, message):
    """
    Defines an elicit slot type response.
    """

    return {
        "sessionAttributes": session_attributes,
        "dialogAction": {
            "type": "ElicitSlot",
            "intentName": intent_name,
            "slots": slots,
            "slotToElicit": slot_to_elicit,
            "message": message,
        },
    }


def delegate(session_attributes, slots):
    """
    Defines a delegate slot type response.
    """

    return {
        "sessionAttributes": session_attributes,
        "dialogAction": {"type": "Delegate", "slots": slots},
    }


def close(session_attributes, fulfillment_state, message):
    """
    Defines a close slot type response.
    """

    response = {
        "sessionAttributes": session_attributes,
        "dialogAction": {
            "type": "Close",
            "fulfillmentState": fulfillment_state,
            "message": message,
        },
    }

    return response

# Validates the data provided by the user.
def validate_data(age, investment_amount):
    
    if age is not None:
        age = int(age)
        if age <= 0 or age > 65:
            return build_validation_result(
                False,
                "age",
                "The age should be betwen 0 and 65, please enter an age again.",
            )
    
    if investment_amount is not None:
        investment_amount = int(investment_amount)
        if investment_amount < 5000:
            return build_validation_result(
                False,
                "investmentAmount",
                "The minimum investment is $5000, please enter an amount again.",
            )
    
    return build_validation_result(True, None, None,)

### Intents Handlers ###
def recommend_portfolio(intent_request):
    """
    Performs dialog management and fulfillment for recommending a portfolio.
    """

    first_name = get_slots(intent_request)["firstName"]
    age = get_slots(intent_request)["age"]
    investment_amount = get_slots(intent_request)["investmentAmount"]
    risk_level = get_slots(intent_request)["riskLevel"]
    source = intent_request["invocationSource"]

    if source == "DialogCodeHook":
        # This code performs basic validation on the supplied input slots.
        slots = get_slots(intent_request)
        
        validation_result = validate_data(age, investment_amount)
        
        # If the data provided by the user is not valid,
        # the elicitSlot dialog action is used to re-prompt for the first violation detected.
        if validation_result['isValid'] == False:
            return elicit_slot(
                intent_request["sessionAttributes"],
                intent_request["currentIntent"]["name"],
                slots,
                validation_result["violatedSlot"],
                validation_result["message"],
            )

        # Fetch current session attibutes
        output_session_attributes = intent_request["sessionAttributes"]

        return delegate(output_session_attributes, get_slots(intent_request))
    
    # Initialize the recommendation variable
    risk_portfolio = ""
    
    # Select portfolio
    if risk_level == "None":
        risk_portfolio = "100% bonds (AGG), 0% equities (SPY)"
        
    elif risk_level == "Very Low":
        risk_portfolio = "80% bonds (AGG), 20% equities (SPY)"
        
    elif risk_level == "Low":
        risk_portfolio = "60% bonds (AGG), 40% equities (SPY)"
        
    elif risk_level == "Medium":
        risk_portfolio = "40% bonds (AGG), 60% equities (SPY)"
        
    elif risk_level == "High":
        risk_portfolio = "20% bonds (AGG), 80% equities (SPY)"

    # Return a message with the initial recommendation based on the risk level.
    return close(
        intent_request["sessionAttributes"],
        "Fulfilled",
        {
            "contentType": "PlainText",
            "content": """Thank you for your information, {}.
            We recommend you an investment portfolio: {}
            """.format(first_name, risk_portfolio),
        },
    )

### Intents Dispatcher ###
def dispatch(intent_request):
    """
    Called when the user specifies an intent for this bot.
    """

    intent_name = intent_request["currentIntent"]["name"]

    # Dispatch to bot's intent handlers
    if intent_name == "recommendPortfolio":
        return recommend_portfolio(intent_request)

    raise Exception("Intent with name " + intent_name + " not supported")


### Main Handler ###
def lambda_handler(event, context):
    """
    Route the incoming request based on intent.
    The JSON body of the request is provided in the event slot.
    """

    return dispatch(event)